esmvalcore/preprocessor/_shared.py

```diff
@@ -230,11 +251,17 @@ def preserve_float_dtype(func: Callable) -> Callable:
     to give output with any type.
 
     """
+    signature = inspect.signature(func)
+    if not signature.parameters:
+        raise TypeError(
+            f"Cannot preserve float dtype during function '{func.__name__}, "
+            f"function takes no arguments"
+        )
```

Comment:
Would it be possible to raise a similar exception if the first argument or the return value does not have a `dtype` attribute? That would make it more clear what is going wrong to preprocessor developers who copy/paste the decorator without knowing what it does.

Reply:
that's a good point there!

Reply:
Do you mean checking the potential type hint of the first argument or return type? Otherwise, this can only be checked when the function is actually run, and not here. In this case, the error `AttributeError: 'int' object has no attribute 'dtype'` is raised right now, which IMHO is already pretty informative. I can of course expand that error message if necessary.

Reply:
Having at least the name of the wrapped function in the error message would be helpful, so if a wrong argument is passed into the function somewhere in people's code, they have some clue where to search. Alternatively, you could just make the `preserve_float_dtype` wrapper a no-op if either the first argument or return value do not have a dtype attribute, leaving error handling to further down the line.